File: backend/src/main/java/de/wps/radvis/backend/fahrradroute/schnittstelle/view/FahrradrouteListenView.java
/*
 * Copyright (c) 2023 WPS - Workplace Solutions GmbH
 *
 * Licensed under the EUPL, Version 1.2 or as soon they will be approved by the European Commission - subsequent versions of the EUPL (the "Licence");
 *
 * You may not use this work except in compliance with the Licence.
 * You may obtain a copy of the Licence at:
 *
 * https://joinup.ec.europa.eu/collection/eupl/eupl-text-eupl-12
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the Licence is distributed on an "AS IS" basis, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the Licence for the specific language governing permissions and limitations under the Licence.
 */

package de.wps.radvis.backend.fahrradroute.schnittstelle.view;

import java.util.Locale;

import de.wps.radvis.backend.fahrradroute.domain.dbView.FahrradrouteListenDbOhneGeomView;
import de.wps.radvis.backend.fahrradroute.domain.valueObject.FahrradrouteName;
import de.wps.radvis.backend.fahrradroute.domain.valueObject.FahrradrouteTyp;
import de.wps.radvis.backend.fahrradroute.domain.valueObject.Kategorie;
import jakarta.annotation.Nullable;
import lombok.Getter;

@Getter
public class FahrradrouteListenView {
	private final Long id;
	private final FahrradrouteName name;
	private final Kategorie fahrradrouteKategorie;
	private final FahrradrouteTyp fahrradrouteTyp;
	@Nullable
	private final String verantwortlicheOrganisation;

	@Nullable
	private String anstiegAbstieg;

	public FahrradrouteListenView(FahrradrouteListenDbOhneGeomView dbView) {
		this.id = dbView.getId();
		this.name = dbView.getName();
		this.fahrradrouteKategorie = dbView.getKategorie();
		this.fahrradrouteTyp = dbView.getFahrradrouteTyp();
		this.verantwortlicheOrganisation = dbView.getVerantwortlicheOrganisationName();

		this.anstiegAbstieg = getAnstiegAbstiegText(dbView.getAnstieg(), dbView.getAbstieg());
	}

	private String getAnstiegAbstiegText(Double anstieg, Double abstieg) {
		return String.format("%s m / %s m",
			getNullableDoubleAsIntegerString(anstieg),
			getNullableDoubleAsIntegerString(abstieg));
	}

	private String getNullableDoubleAsIntegerString(Double d) {
		return d == null ? "-" : String.format(Locale.GERMANY, "%d", d.intValue());
	}
}

File: frontend/radvis/src/app/viewer/viewer-shared/models/fahrradroute-listen-view.ts
/*
 * Copyright (c) 2023 WPS - Workplace Solutions GmbH
 *
 * Licensed under the EUPL, Version 1.2 or as soon they will be approved by the European Commission - subsequent versions of the EUPL (the "Licence");
 *
 * You may not use this work except in compliance with the Licence.
 * You may obtain a copy of the Licence at:
 *
 * https://joinup.ec.europa.eu/collection/eupl/eupl-text-eupl-12
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the Licence is distributed on an "AS IS" basis, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the Licence for the specific language governing permissions and limitations under the Licence.
 */

import { FahrradrouteKategorie } from 'src/app/viewer/viewer-shared/models/fahrradroute-kategorie';
import { FahrradrouteTyp } from 'src/app/viewer/viewer-shared/models/fahrradroute-typ';

export interface FahrradrouteListenView {
  id: number;
  fahrradrouteTyp: FahrradrouteTyp;
  name: string;
  fahrradrouteKategorie: FahrradrouteKategorie;
  verantwortlicheOrganisation: string;

  anstiegAbstieg: string;
}

// eslint-disable-next-line @typescript-eslint/no-namespace
export namespace FahrradrouteListenView {
  const EMPTY_FIELD_INDICATOR = '';

  export const getDisplayValueForKey = (item: FahrradrouteListenView, key: string): string | string[] => {
    switch (key) {
      case 'name':
        return item.name ?? EMPTY_FIELD_INDICATOR;
      case 'verantwortlicheOrganisation':
        return item.verantwortlicheOrganisation ?? EMPTY_FIELD_INDICATOR;
      case 'fahrradrouteKategorie':
        return (
          FahrradrouteKategorie.options.find(option => option.name === item.fahrradrouteKategorie)?.displayText ??
          EMPTY_FIELD_INDICATOR
        );
      case 'fahrradrouteTyp':
        return (
          FahrradrouteTyp.options.find(option => option.name === item.fahrradrouteTyp)?.displayText ??
          EMPTY_FIELD_INDICATOR
        );
      case 'anstiegAbstieg':
        return item.anstiegAbstieg ?? EMPTY_FIELD_INDICATOR;
      default:
        throw Error(`Key ${key} nicht gefunden`);
    }
  };
  export const getDisplayName = (fahrradrouteListenView: FahrradrouteListenView | null | undefined): string => {
    if (!fahrradrouteListenView?.name) {
      return '';
    }

    return fahrradrouteListenView.name;
  };
}
